### PR TITLE
Hot Fix - Fix potential inconsistent w/v tensor list length in distributed checkpoint loading when DP changes

### PR DESCRIPTION
**Pull request by yzygitzh**
In legacy grouped GeMM with GLU, experts need to be split along dim-0 for DP sharding, and along dim-2 for GLU.

The following sketch, which is from the function comment of `sh_ten_build_fn` in `megatron/core/transformer/moe/experts.py`, shows a case of DP sharding for expert weight:

```
|................|           |........|........|
|............FFFF|           |........|....BBBB|
|FFFFFFFFFFFFFFFF|     ->    |AAAAAAAA|BBBBBBBB|
|FFFFFFFFFFFFFFFF|           |AAAAAAAA|BBBBBBBB|
|FF..............|           |AA......|........|
|................|           |........|........|
```

There are 3 A-chunks (for W) and 3 B-chunks (for V). However, the number of A/B chunks can be non-equal in some sharding conditions, and the final `torch.cat` in `sh_ten_merge_fn` can fail due to not accepting `None` elements padded by previous `zip_longest`.

This commit pads w/v tensors with empty tensors instead of None for such cases.

### CODE REVIEW

**Review by sjtu-yc — APPROVED** [review]

**Review by lx865712528 — APPROVED** [review]

File: megatron/core/transformer/moe/experts.py [hunk 1]
```diff
@@ -563,13 +563,18 @@ def sh_ten_merge_fn(sub_state_dict, tp_axis: int, with_glu: bool):
                         local_expert_dict['v_tensors'], local_expert_dict['v_lens']
                     )
                     first_glu_idx = local_expert_dict['first_glu_idx']
+                    fill_tensor = None
+                    if len(w_tensors):
+                        fill_tensor = w_tensors[0].new_empty(0)
+                    elif len(v_tensors):
```
Comment (abuccts) [inline megatron/core/transformer/moe/experts.py:569]: what if both lens > 0, but len(w) != len(v)?

**Review by abuccts — APPROVED** [review]

File: megatron/core/transformer/moe/experts.py (hunk 1 above)
Comment (yzygitzh) [reply megatron/core/transformer/moe/experts.py:569]: Keep as it is per offline discussion.

File: megatron/core/transformer/moe/experts.py [hunk 2]
```diff
@@ -563,13 +563,18 @@ def sh_ten_merge_fn(sub_state_dict, tp_axis: int, with_glu: bool):
...
+                        fill_tensor = v_tensors[0].new_empty(0)
```
[the remark this answers is not in the document]
Comment (yzygitzh) [reply megatron/core/transformer/moe/experts.py:570]: If both w_tensors and v_tensors are empty, the following zip_longest will crash. No need to add extra check.
